chapters/3/modul-kompilujacy.tex

```diff
@@ -0,0 +1,75 @@
+\begin{verbatim}
+    /folder powiązany
+    |-- /source
+    |   |-- main.cpp
+    |-- /input
+    |   |-- 1.in
+    |   |-- 2.in
+    |   |-- n.in
+    |-- /output
+        |-- 1.out
+        |-- 2.out
+        |-- n.out
+\end{verbatim}
+W katalogu /source znajdują się pliki zawierające kod źródłowy c++. Wejście, które chcemy przekazać przy uruchamianiu programu znajduje się w plikach *.in w folderze /input. W trakcie działania programu tworzone są pliki wynikowe z rozszerzeniem *.out w folderze /output. 
+
+\subsection{Skrypt uruchamiany przy starcie kontenera}
+Skrypt entrypoint to skrypt wskazany w pliku Dockerfile, który uruchamia się po starcie kontenera. Został napisany w Bashu i realizuje funkcjonalność modułu kompilującego, kompiluje pliki *.cpp znajdujące się w katalogu /source powiązanego folderu za pomocą mingw, a następnie uruchamia skompilowany program dla każdego z plików wejścia *.in umieszczonych w folderze /in. Wynik każdego uruchomienia przekierowywany jest do  pliku wynikowego *.out w folderze /out katalogu współdzielonego. Każde uruchomienie dodatkowo mierzy czas wykonania przy pomocy komendy date.
+\begin{figure}[!h]
+	\begin{center}
+		\resizebox{1\textwidth}{!} {
+			\includegraphics{img/3/diagram-aktywnosci-modul-kompilujacy.png}
+		}
+		\caption[Diagram aktywności modułu kompilującego]{Diagram wizualizujący przebieg pracy modułu kompilującego. Źródło własne.}
+		\label{diagram-aktywnosci-worker}
+	\end{center}
+\end{figure}
+
+\subsection{Ogranienie zasobów kontenera}
+Jednym z elementów zapewnienia sprawiedliwej oceny czasu wykonania zadania jest określenie maksymalnej ilości pamięci ram oraz zasobów procesora, które są dostępne dla kontenera. Domyślnie kontener nie posiada żadnych ograniczeń wykorzystywanych zasobów i wykorzystuje ich tyle, ile zostanie mu przydzielone przez jądro systemu hosta. Te wartości można jednak ograniczyć, korzystając z mechanizmu kontroli maksymalnych zasobów wykorzystanych przez kontener udostępnionego przez narzędzie Docker\cite{dockerConstraints}, poprzez użycie flag --memory oraz --cpus przy starcie kontenera. Zakładając, że przydzielone kontenerom zasoby nie przekraczają całości dostępnych zasobów komputera, możemy osiągnąć względnie sprawiedliwą ocenę czasu wykonania zadania. Rozwiązaniem, które pozwala z większą dokładnością określić zasoby czasowe potrzebne do wykonania programu jest pomiar wykorzystania procesora na podstawie liczby cykli, jednak implementacja tej funkcjonalności wewnątrz kontenera dockerowego wiąże się z koniecznością uruchomienia go ze zwiększonymi uprawnieniami do funkcji jądra, co znacząco zmniejsza izolację modułu i bezpieczeństwo całego systemu, w szczególności biorąc pod uwagę fakt, że moduł ten odpowiada za kompilację i wykonanie potencjalnie szkodliwego kodu przesłanego przez użytkowników zewnętrznego systemu STOS.
```

Comment:
Ja u siebie coś podobnego bardzo napisałem xd
Ale to później ustalimy czy wywalić